src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorDiagnosticDescriptor.cs

```diff
@@ -1,75 +1,33 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
 using System.Diagnostics;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-[DebuggerDisplay("{" + nameof(DebuggerToString) + "(),nq}")]
-public sealed class RazorDiagnosticDescriptor : IEquatable<RazorDiagnosticDescriptor>
+[DebuggerDisplay($"{{{nameof(DebuggerToString)}(),nq}}")]
+public sealed record RazorDiagnosticDescriptor
 {
-    private readonly Func<string> _messageFormat;
+    public string Id { get; }
+    public string MessageFormat { get; }
+    public RazorDiagnosticSeverity Severity { get; }
 
-    public RazorDiagnosticDescriptor(
-        string id,
-        Func<string> messageFormat,
-        RazorDiagnosticSeverity severity)
+    public RazorDiagnosticDescriptor(string id, string messageFormat, RazorDiagnosticSeverity severity)
```

Comment:
```suggestion
    public RazorDiagnosticDescriptor(string id, string? messageFormat, RazorDiagnosticSeverity severity)
```

Reply:
FWIW, I'm inclined to leave this as non-nullable. While we have a fallback for null and tests that ensure that the fallback works, that's not the intent of the API and it's not how it's used in the compiler. Nowhere in the compiler do we actually use this fallback.

I suspect that the fallback was added for `RazorDiagnostic` descendants that behave badly, since it used to be possible for a third-party to inherit from this class and provide their own diagnostics. So, I would be more inclined to remove the fallback completely along with the tests that are you noted below. What do you think?

Reply:
I would also be inclined to remove the fallback. Creating a diagnostic descriptor without a message to format is near-certainly an error.